### DISCUSSION TITLE
Flow Chart of min_cost_flow

### DISCUSSION THREAD

**Discussion by ayliniremkara** [networkx/networkx#5546] [Q&A]
Hello team,

Is there any process flow chart for min_cost_flow?
https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.flow.min_cost_flow.html

**Comment by rossbar** [comment]
> Is there any process flow chart for min_cost_flow?

Could you provide a bit more info? Do you mean some sort of visualization of the algorithm in the docstring?

**Reply by ayliniremkara** [reply]
Thank you for your answer, yes could you please share it with me if it exists?

**Reply by MridulS** [reply]
We currently do not have anything that visually explains the algorithm, but this would definitely be something we would be interested in :)

**Reply by stanyas** [reply]
Here is an example of a visualization:

![figure1](https://user-images.githubusercontent.com/8345042/167161293-ee66ba6b-6c72-4164-a929-75e21fb3c15d.png)

f is the flow from the node, w is the weight per unit (cost), and c is the capacity of the edge to bear the transportation. Total minimal cost flow cost is f*w summed through all edges, 34. The demand needs to sum up to 0. In this case, demand of a and c was -5 (it needed to give 5) and b and d was 5 (it needed to get 5).

Here is the code for generating it, altered from [the docs example](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.flow.min_cost_flow.html):
```

import numpy as np
import matplotlib.pyplot as plt
import networkx as nx

G = nx.DiGraph()

G.add_node("a", demand=-5)
G.add_node("d", demand=5)
G.add_node("c", demand=-5)
G.add_node("b", demand=5)

G.add_edge("a", "b", weight=3, capacity=4)
G.add_edge("a", "c", weight=6, capacity=10)
G.add_edge("b", "d", weight=1, capacity=9)
G.add_edge("c", "d", weight=2, capacity=5)
G.add_edge("c", "b", weight=6, capacity=1)

flowDict = nx.min_cost_flow(G)
flowCost = nx.min_cost_flow_cost(G)

pos = nx.spring_layout(G)

nx.draw_networkx_edge_labels(
    G,
    pos,
    edge_labels={
        ("a", "b"): "f= " + str(flowDict["a"]["b"]) + ", w=3, c=4",
        ("a", "c"): "f= " + str(flowDict["a"]["c"]) + ", w=6, c=10",
        ("b", "d"): "f= " + str(flowDict["b"]["d"]) + ", w=1, c=9",
        ("c", "d"): "f= " + str(flowDict["c"]["d"]) + ", w=2, c=5",
        ("c", "b"): "f= " + str(flowDict["c"]["b"]) + ", w=6, c=1",
    },
    font_color="black",
)

nx.draw(
    G,
    pos,
    with_labels=True,
    node_size=5,
    font_size=7,
    edge_color="red",
    node_color="tab:red",
)

print(flowCost)
plt.show()

```
I am not sure though if this was what you wanted. Perhaps you meant how the algorithm works instead?

**Comment by rossbar** [comment]
It's not *exactly* the algorithm you're looking for, but I did want to point out that there is a very nice detailed explanation of the Dinitz algorithm for max flow problem in [nx-guides](https://networkx.org/nx-guides/content/algorithms/flow/dinitz_alg.html).

@ayliniremkara I'm curious if something like this is what you had in mind, or whether you were looking something more like the example above, or something else entirely...
